### PR TITLE
[8.x] Update publicPath method, update docblocks, add tests

### PR DESCRIPTION
**Pull request by iamgergo**
This PR makes the `App::publicPath` parameterizable, cleans up some docblocks, and adds some tests to the path generator methods.

Until now, we had to use the `public_path` helper, when generating paths to the public directory. It's not a big deal, but in some cases we might use the `App` facade instead of the helper, and it feels a bit mixed-up:

```php
$this->publishes([
    __DIR__.'/../public' => public_path('vendor/bazar'),
    __DIR__.'/../resources/js' => App::resourcePath('js/vendor/bazar'),
    __DIR__.'/../resources/sass' => App::resourcePath('sass/vendor/bazar'),
], 'bazar-assets');
```

This is not a BC since the `publicPath` method is not present in the `Application` contract.

> The `storatePath` is present, so changing that would be a breaking change I think.

### CODE REVIEW

File: src/Illuminate/Foundation/Application.php [hunk 1]
```diff
@@ -443,11 +443,12 @@ public function useLangPath($path)
     /**
      * Get the path to the public / web directory.
      *
+     * @param  string  $path
      * @return string
      */
-    public function publicPath()
+    public function publicPath($path = '')
```
Comment (driesvints) [inline src/Illuminate/Foundation/Application.php:449]: This is a breaking change to the method signature unfortunately.

**Comment by driesvints** [comment]
Seems like this is breaking the test suite. Feel free to re-attempt this with passing tests and no breaking changes.

**Comment by iamgergo** [comment]
Well, I don't think it's a breaking change, the Mix tests that are chaning the path totally:

```php
        $this->app->singleton('path.public', function () {
            return __DIR__;
        });
```

**Comment by driesvints** [comment]
@iamgergo changing a method signature is a breaking change for anyone overriding it.

**Comment by iamgergo** [comment]
Yes, in that case that's right. I was considering only interfaces.

**Pull request closed without merging**